### ISSUE TITLE
Achievement Chain

### ISSUE THREAD

**Issue by gabriel-simonetti** [gstt/laravel-achievements#30]
Sometimes, there are multiple achievements that track the exact same thing, but with different points to obtain. 

For example, an user can obtain a "Newcomer" achievement when creating their first comment on the site, a "Regular Commenter" achievement when they reached 10 comments and a "Regular Commenter" achievement when they reached 100 comments.

Currently, these achievements need to be tracked separately, something like this:

```php
<?php
use App\Achievements\Newcomer;
use App\Achievements\RegularCommenter;
use App\Achievements\VeteranCommenter;

public function createComment($user, $comment){
     // Some logic to be processed when the user creates a comment...
    $user->addProgress(Newcomer, 1);
    $user->addProgress(RegularCommenter, 1);
    $user->addProgress(VeteranCommenter, 1);
}
```

And this would quickly grow out of hand when there are many achievements tracking the same thing. In order to solve this problem, I propose a feature called "Achievement Chain". It would work in the following way:

1) A new abstract class, `AchievementChain` will be created. This class will be defined only by a list of achievements.
2) All methods that can be used to add, remove, set reset progress to Achievements can also receive instances of AchievementChain. When one of this methods receive an instance of AchievementChain, it will call the same method for every Achievement described in the definition of the AchievementChain instance.

**Comment by kgenly** [comment]
YES! I came here looking for exactly this functionality. I have some workarounds going in the meanwhile, but I think this is a grand solution.

Another potential thing to look at is whether an achievement is "replaceable" - that is, if an achievement higher in the chain has been unlocked, the lower ones stop displaying. Thereby "transforming" my "newcomer" achievement into a "regular commenter" achievement.

**Comment by gabriel-simonetti** [comment]
In the way you put it, "Replaceable" seems like a good idea but I think it's outside of the scope of this project - since it's more about how achievements are displayed on the frontend instead of how they're store on the backend.

But I agree that sometimes it's necessary to retrieve the highest achievement unlocked in the chain. Maybe the AchievementChain could return an instance of AchievementProgress of the higher achievement unlocked in the chain? I'm thinking of something like:

```php
<?php
use App\Achievements\Newcomer; // Create 1 comment
use App\Achievements\RegularCommenter; // Create 10 comments
use App\Achievements\VeteranCommenter; // Create 100 comments

class CommentAchievements implements AchievementChain{
    public function chain(){
        return [Newcomer::class, RegularCommenter::class, VeteranCommenter::class];
    }
}
```
```
// $user1 has unlocked RegularCommenter
// $user2 has unlocked VeteranCommenter
$user1->highestOnAchievementChain(CommentAchievements); // returns instance of AchievementProgress for RegularCommenter
$user2->highestOnAchievementChain(CommentAchievements); // returns instance of AchievementProgress for VeteranCommenter
```

**Comment by JoNMii** [comment]
> The feature for AchievementChain has been pushed to the master branch. It's still on development since there is no release version for it.

Is the command "php artisan achievements:load" working with chains for you?
